jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java

```diff
@@ -209,38 +218,150 @@ private static StringBuilder preprocessLayout(MemoryLayout targetLayout, boolean
-				if (isPaddingLayout(structElement)) {
-					paddingElements += 1;
-					/* The padding bytes is required in the native signature for upcall thunk generation. */
-					if (!isDownCall) {
-						elementLayoutStrs.append('(').append(structElement.byteSize()).append(')');
-					}
-				} else {
-					elementLayoutStrs.append(preprocessLayout(structElement, isDownCall));
+		/*[IF JAVA_SPEC_VERSION >= 21]*/
+		} else if (targetLayout instanceof UnionLayout unionLayout) { /* Intended for the nested union since JDK21. */
+			targetLayoutStr = encodeUnionLayoutStr(unionLayout, targetLayoutStr, isDownCall);
+		/*[ENDIF] JAVA_SPEC_VERSION >= 21 */
+		} else if (targetLayout instanceof GroupLayout groupLayout) { /* Intended for the nested struct. */
+			targetLayoutStr = encodeStructLayoutStr(groupLayout, targetLayoutStr, isDownCall);
+		}
+
+		return targetLayoutStr;
+	}
+
+	/* Encode the types in the struct layout to a symbol string to simplify the further processing in native. */
+	private static StringBuilder encodeStructLayoutStr(GroupLayout structLayout, StringBuilder targetLayoutStr, boolean isDownCall) {
+		List<MemoryLayout> elementLayoutList = structLayout.memberLayouts();
+		int structElementCount = elementLayoutList.size();
+		StringBuilder elementLayoutStrs = new StringBuilder();
+		int paddingElements = 0;
+
+		for (MemoryLayout structElement : elementLayoutList) {
+			/* Ignore any padding element in the struct as it is handled by ffi_call by default. */
+			if (isPaddingLayout(structElement)) {
+				paddingElements += 1;
+				/* The padding bytes is required in the native signature for upcall thunk generation. */
+				if (!isDownCall) {
+					elementLayoutStrs.append('(').append(structElement.byteSize()).append(')');
 				}
+			} else {
+				elementLayoutStrs.append(preprocessLayout(structElement, isDownCall));
 			}
-			/* Prefix "#" to denote the start of this layout string in the case of downcall. */
-			if (isDownCall) {
-				targetLayoutString.append('#').append(structElementCount - paddingElements);
+		}
+
+		/* Prefix "#" to denote the start of this layout string in the case of downcall. */
+		if (isDownCall) {
+			targetLayoutStr.append('#').append(structElementCount - paddingElements);
+		}
+		targetLayoutStr.append('[').append(elementLayoutStrs).append(']');
+
+		return targetLayoutStr;
+	}
+
+	/*[IF JAVA_SPEC_VERSION >= 21]*/
+	/* Encode the types in the union layout to a symbol string to simplify the further processing in native. */
+	private static StringBuilder encodeUnionLayoutStr(UnionLayout unionLayout, StringBuilder targetLayoutStr, boolean isDownCall) {
+		long unionByteSize = unionLayout.byteSize();
+		StringBuilder elementLayoutStrs = new StringBuilder();
+		int byteAlignment = (int)unionLayout.byteAlignment();
+		ValueLayout sequenceElement;
+
+		/* The idea is to select the primitve type with the biggest alignement of the union
```

Comment:
Please correct the spelling of "primitive" here and on line 322.

Reply:
Fixed.

Reply:
On the same line, there's a typo in "alignment".

Reply:
Fixed.